### DISCUSSION TITLE
converting an old project to targets with time-consuming intermediate data objects

### DISCUSSION THREAD

**Discussion by venpopov** [ropensci/targets#1377] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

In the process of learning how to work with targets, I decided to convert an existing ongoing project. I've learned how to do it from scratch, but I have the following "problem". There is a computationally demanding step which simulates some models, which originally took about a week to complete. The object produced by this step is used in downstream analyses. I would like to avoid recomputing this object and just reuse it. Is it somehow possible to setup the targets pipeline such that it will consider some steps completed and reuse the (externally-generated) object as if it was produced by a step in the targets pipeline? I hope my question makes sense.

**Comment by joelnitta** [comment]
You could save the object to an RDS file, then use [tar_file()](https://docs.ropensci.org/tarchetypes/reference/tar_formats.html) to track the file contents and read it in to your pipeline. If for any reason the contents of the RDS file changed it would trigger downstream steps that depend on it. But AFAIK {targets} has no way of knowing how changes *upstream* of the model object would affect it, other then completely re-doing the workflow in {targets}.

**Reply by venpopov** [reply]
That's what I was thinking, although I would have preferred some way to "trick" targets to think this file is an object produced by a step in the pipeline, and still leave the possibility to rerun everything if needed. The best solution I can think of is comment out the relevant target steps that would normally produce this file, so that it is at least clear how it would have come to be.

**Reply by joelnitta** [reply]
Yeah, I don't think you can have it both ways. I think one of the points of using {targets} is that every target should be completely determined by the code. It provides proof that the results are derived from the code. Having an object that appears to be produced by the code but actually isn't kind of defeats the point of the package.

Maybe something like this?


```
make_model <- function(model_object_file) {
  # code that actually made the model, but we don't run
  # ...
  readRDS(model_object_file)
}

list(
  tar_target(model_object_file, "model.RDS", format = "file"),
  tar_target(model, make_model(model_object_file))
)

```

**Reply by venpopov** [reply]
That's a really good justification for why not to allow this behavior, even if it were possible to do :)
I guess I'll just live with the annoyances that inevitably come with any workflow transition, and hope the long-term payoff is worth it.

I like that function, it's cleaner than having commented targets.

**Reply by joelnitta** [reply]
I'm glad it was helpful 🙂

TBH in my experience, it's not that unusual to end up in a situation where you have to re-run a long-running target even if you didn't mean to. You may have inadvertently changed something upstream, or on rare occasions there are updates to {targets} that can't use the old version's cache. You might consider [gittargets](https://github.com/ropensci/gittargets) to version-control the cache, if that is a concern going forward (mind though that this comes with its own mental and logistical overhead).

**Reply by venpopov** [reply]
Yeah, I had exactly that experience 6 months ago when I originally tried to transition to targets. Tried `gittargets` and eventually gave up because of the mental overhead exactly. My existing mental models and workflows - refactoring functions, switching branches for experiments, etc, didn't seem compatible with targets. Eventually what drew me to targets - the promise to avoid recomputing expensive steps - was counteracted. I'm hoping this time it goes better as I do like the idea behind it.

**Reply by joelnitta** [reply]
I suppose the ideal situation would be if {gittargets} wasn't needed as a separate package because that functionality was totally abstracted away and embedded within {targets}: if it saw you were re-running the code in a state that matched something exactly in the past, it could just restore the cache from that previous state. But this probably is much easier said than done!

**Reply by wlandau** [reply]
I may actually supersede `gittargets` in favor of content-addressable storage, c.f. https://github.com/ropensci/gittargets/issues/18
